cloudinit/config/cc_reset_rmc.py

```diff
@@ -0,0 +1,138 @@
+# (c) Copyright IBM Corp. 2020 All Rights Reserved
+#
+# Author: Aman Kumar Sinha <amansi26@in.ibm.com>
+#
+# This file is part of cloud-init. See LICENSE file for license information.
+
+
+"""
+Reset RMC
+------------
+**Summary:** reset rsct node id
+
+Reset RMC module is IBM PowerVM Hypervisor specific
+
+Reliable Scalable Cluster Technology (RSCT) is a set of software components,
+that  together provide a comprehensive clustering environment (RAS features)
+for IBM PowerVM based virtual machines. RSCT includes the Resource monitoring
+and control (RMC) subsystem. RMC is a generalized framework used for managing,
+monitoring, and manipulating resources. RMC runs as a daemon process on
+individual machines and needs creation of unique node id and restarts
+during VM boot.
+More details refer
+https://www.ibm.com/support/knowledgecenter/en/SGVKBA_3.2/admin/bl503_ovrv.htm
+
+This module handles
+- creation of the unique RSCT node id to every instance/virtual machine
+  and ensure once set, it isn't changed subsequently by cloud-init.
+  In order to do so, it restarts RSCT service.
+
+Prerequisite of using this module is to install RSCT packages.
+
+**Internal name:** ``cc_reset_rmc``
+
+**Module frequency:** per instance
+
+**Supported distros:** rhel, sles and ubuntu
+
+"""
+import os
+
+from cloudinit import log as logging
+from cloudinit.settings import PER_INSTANCE
+from cloudinit import util
+from cloudinit import subp
+
+frequency = PER_INSTANCE
+PIDOF = 'pidof'
+
+# RMCCTRL is expected to be in system PATH (/opt/rsct/bin)
+# The symlink for RMCCTRL and RECFGCT are
+# /usr/sbin/rsct/bin/rmcctrl and
+# /usr/sbin/rsct/install/bin/recfgct respectively.
+RSCT_PATH = '/opt/rsct/install/bin'
+RMCCTRL = 'rmcctrl'
+RECFGCT = 'recfgct'
+
+LOG = logging.getLogger(__name__)
+
+NODE_ID_FILE = '/etc/ct_node_id'
+
+
+def handle(name, _cfg, cloud, _log, _args):
+    # Ensuring node id has to be generated only once during first boot
+    if cloud.datasource.platform_type == 'none':
+        LOG.debug('Skipping creation of new ct_node_id node')
+        return
+
+    orig_path = os.environ.get('PATH')
+    try:
```

Comment:
the reason that i suggested this code block (lines 72-95) be moved to a function is so that it is easier to unit test and also to read.  

 * to unit test: then you can just call that function... you don't need the difficult setup required to call handle().  As I've said before, the pain in calling handle is *my fault* for designing it that way. sorry :cry: . To avoid the pain, the general solution is just to minimize what handle actually does.

 * to read: you can look at a code block like this and say "oh, all that does is call a function with a modified PATH and then set it back".   It also reduces indentation by one level.

```python
try:
    add_path(orig_path)
    
    call_your_function()
finally:
    if orig_path:
        os.environ['PATH'] = orig_path
    else:
        del os.environ['PATH']
```


if we had a context handler, it reads even better, but writing one for one
caller doesn't seem necessary.  Then it might look like this:

```python

with prepended_path(RSCT_PATH):
    call_your_function()
```

Reply:
I think I get your point.
Using handle in UT it is vary hard to handle datasource in UTs. This is much easier in running UT.

Reply:
If you're interested in using it, I was interested in trying, and wrote 'prepended_path' context manager.


```python
import contextlib
import os

@contextlib.contextmanager
def prepended_path(path):
    enter_path = os.environ.get("PATH")
    new_path = path + ((":" + enter_path) if enter_path else "")
    try:
        os.environ["PATH"] = new_path
        yield new_path
    finally:
        assert os.environ.get("PATH") == new_path, (
            "PATH changed during prepended_path context "
            "old=%s new=%s" % (enter_path, new_path))

        if enter_path is None:
            del os.environ["PATH"]
        else:
            os.environ["PATH"] = enter_path


print("before: %s" % os.environ.get("PATH"))

with prepended_path("/mybin"):
    print("during: %s" % os.environ.get("PATH"))

print("after: %s" % os.environ.get("PATH"))
```

So you could use this just as:

```python

with prepended_path(RSCT_PATH):
    reset_rmc()
```

I think that reads really nicely.